python/cudf/cudf/core/join/join.py

```diff
@@ -0,0 +1,431 @@
+        return libcudf_join_type
+
+    def libcudf_to_output_casting_rules(self, lcol, rcol, how):
+
+        dtype_l = lcol.dtype
+        dtype_r = rcol.dtype
+
+        merge_return_type = None
+        # we  currently only need to do this for categorical variables
+        if (is_categorical_dtype(dtype_l) or is_categorical_dtype(dtype_r)):
+            if how in ['inner', 'left']:
+                merge_return_type = dtype_l
+            elif how == 'outer':
+                new_cats = cudf.concat(dtype_l.categories, dtype_r.categories)
+                merge_return_type = cudf.core.dtypes.CategoricalDtype(categories=new_cats)
+        return merge_return_type
+
+    def compute_output_dtypes(self):
+        index_dtypes = []
+        data_dtypes = {}
+        for name, col in itertools.chain(self.lhs._data.items(), self.rhs._data.items()):
+                data_dtypes[name] = col.dtype
+
+        if self.left_index and self.right_index:
+            for ((kl, vl), (kr, vr)) in zip(
+                self.lhs.index._data.items(),
+                self.rhs.index._data.items()
+            ):
+                to_dtype = self.libcudf_to_output_casting_rules(vl, vr, self.how)
+                index_dtypes.append(to_dtype)
+            if self.left_on and self.right_on:
+                for ((kl, vl), (kr, vr)) in zip(
+                    self.lhs[self.left_on]._data.items(),
+                    self.rhs[self.right_on]._data.items()
+                ):
+                    to_dtype = self.libcudf_to_output_casting_rules(vl, vr, self.how)
+                    data_dtypes[kl] = to_dtype
+                    data_dtypes[kr] = to_dtype
+        elif self.left_index and self.right_on:
+            for ((kl, vl), (kr, vr)) in zip(
+                self.lhs.index._data.items(),
+                self.rhs[self.right_on]._data.items()
+            ):
+                to_dtype = self.libcudf_to_output_casting_rules(vl, vr, self.how)
+                data_dtypes[kr] = to_dtype
+                index_dtypes = [col.dtype for col in self.rhs.index._data.values()]
+        elif self.right_index and self.left_on:
+            for ((kl, vl), (kr, vr)) in zip(
+                self.lhs[self.left_on]._data.items(),
+                self.rhs.index._data.items()
+            ):
+                to_dtype = self.libcudf_to_output_casting_rules(vl, vr, self.how)
+                data_dtypes[kl] = to_dtype
+                index_dtypes = [col.dtype for col in self.lhs.index._data.values()]
+        elif self.left_on and self.right_on:
+            for ((kl, vl), (kr, vr)) in zip(
+                self.lhs[self.left_on]._data.items(),
+                self.rhs[self.right_on]._data.items()
+            ):
+                to_dtype = self.libcudf_to_output_casting_rules(vl, vr, self.how)
+                data_dtypes[kl] = to_dtype
+                data_dtypes[kr] = to_dtype
+        return (index_dtypes, data_dtypes)
```

Comment:
Similar as above.

Reply:
I didn't manage to simplify this quite as much as the input function. Part of the problem is that depending on the merge parameters we actually need to do somewhat distinctly different things. It seems that any way we factor out the loop, we end up needing to re-check the merge parameters later to decide if the datatypes we've collected correspond to an index, data columns etc. 

This is my least favorite part of the code and I'm open to suggestions of ways to rework it completely.

Reply:
I figured out a way the factor out the loops here, better logic despite similar LOC. Should be updated.